### ISSUE TITLE
🚀 Feature: Remove the `ComponentType` type where possible

### ISSUE THREAD

**Issue by schultzp2020** [backstage/backstage#19705]
### 🔖 Feature description

Remove the `ComponentType` type where possible.

### 🎤 Context

According to [ADR006](https://backstage.io/docs/architecture-decisions/adrs-adr006) backstage wants to avoid use of the types `FC` and `SFC`; however, `ComponentType` is a superset of the two types. Therefore, all references to the `ComponentType` type should be removed where possible. 

### ✌️ Possible Implementation

Backstage should switch to the `(props: ...) => JSX.Element` type because it is what most codebases use with the additional benefit of better intellisense and stricter types. This will also help with the React 18 migration. 

### 👀 Have you spent some time to check if this feature request has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

Yes I am willing to submit a PR!

**Comment by freben** [comment]
> `ComponentType` is a superset of the two types. Therefore, all references to the `ComponentType` type should be removed where possible

I don't think we should do that at this point. React 18 addressed the `children` issue, and I think that going forward we may want to keep using that type where it makes sense, especially as an input type.

**Comment by schultzp2020** [comment]
Hello @freben 

There are two primary rationales for discontinuing the use of the `React.ComponentType` type within our codebase. Firstly, this decision aligns with our architectural choice to remove `React.FunctionalComponent` from our codebase, as using `React.ComponentType` is just a superset of `React.FunctionalComponent`:

```tsx
type ComponentType<P = {}> = ComponentClass<P> | FunctionComponent<P>;
```

Secondly, it's recommended to type functional components by annotating the return type with `JSX.Element`. With the latest React 18 types and TypeScript 5.1, opting for `React.FunctionalComponent` is mostly a matter of style, while the use of `React.FunctionalComponent` is discouraged.

For more information please [read](https://react-typescript-cheatsheet.netlify.app/docs/basic/getting-started/function_components/).

**Comment by freben** [comment]
Right, so it agrees that it's mostly a stylistic choice for recent TS+React.

For me personally, I'm happy to CREATE components with implicit types. But does it get a bit more nuanced for INPUT types? As in, places that accept a component as input. What should the standard signature be there?

**Comment by schultzp2020** [comment]
The explicit vs implicit return type debate is not really a debate when it comes down to providing a library for users for the following reasons.

Pros of explicit return types
1. Understanding the intent of a function.
2. Easier to catch yourself doing something strange.
3. It helps you reason early about your types and make more safe typeguards. 
4. Faster transpile time. 

How using explicit return types compounds in a library
1. Contributors can understand the codebase better.
2. Contributors can contribute to the codebase without fear of accidently returning the wrong type.
3. Contributors are less likely to accidently change the return type for downstream users.
4. Faster transpile time.

Here is a good [video](https://www.youtube.com/watch?v=Gcr4t6cH-lU) going more in depth on why everyone should always use explicit return types. 

The input type for a React component are a separate discussion. You should always use the type `ReactNode` here. There are better ways to handle the typing of a React component than `ComponentType`. I would be happy to discuss this further if you would like. Ideally through a call or something since it's a rather long discussion. Using `ComponentType` also adds unnecessary properties such as `defaultProps` and `displayName`.

**Comment by schultzp2020** [comment]
@freben I would suggest reading these two articles on React prop types, they are very insightful. 

[React.ReactNode vs JSX.Element vs React.ReactElement](https://www.totaltypescript.com/jsx-element-vs-react-reactnode)
[How to Pass a Component as a Prop in React](https://www.totaltypescript.com/pass-component-as-prop-react)

**Comment by freben** [comment]
The first one says

> I'd also suggest not annotating the return types of your components to avoid confusion - but if you're using TypeScript 5.1, go ahead

and also

> In TypeScript versions before 5.1, you can't use React.ReactNode in one specific case - typing the return type of a component.

So does that mean that we need to hold off on that?

**Comment by schultzp2020** [comment]
@freben 

> The first one says

> > I'd also suggest not annotating the return types of your components to avoid confusion - but if you're using TypeScript 5.1, go ahead

Everything should have explicit returns in a library that is meant to be consumed from what I have read. Please [read](https://github.com/backstage/backstage/issues/19705#issuecomment-1706877163) my explanation on "How using explicit return types compounds in a library." I am a bit baffled why there is so much hesitancy in explicit return types. Explicitly declaring return types generally takes only a few seconds anyways. Further, TypeScript is meant to be a type safe language, lets use that power to our advantage. 

> and also

> > In TypeScript versions before 5.1, you can't use React.ReactNode in one specific case - typing the return type of a component.

> So does that mean that we need to hold off on that? Especially for external-facing interfaces since we don't force anybody to pick a particularly recent TS version as consumers of our libraries.

Yes, at least until TypeScript 5.4 is released because of the [TypeScript Releases Policy](https://backstage.io/docs/overview/versioning-policy/#typescript-releases). In the meantime we can either use `JSX.Element` or `React.ReactElement`. I decided to stick with `JSX.Element` since the codebase is already using it. Ideally, I think if we want to change the return type we should do that in another PR.

**Comment by freben** [comment]
> Everything should have explicit returns in a library that is meant to be consumed from what I have read. Please [read](https://github.com/backstage/backstage/issues/19705#issuecomment-1706877163) my explanation on "How using explicit return types compounds in a library." I am a bit baffled why there is so much hesitancy in explicit return types. Explicitly declaring return types generally takes only a few seconds anyways. Further, TypeScript is meant to be a type safe language, lets use that power to our advantage.

Oh there really isn't any hesitancy per se. Everyone is perfectly free to add explicit return types to make things better. This is just me pondering how prescriptive or forcing we should be, on the sliding scale between "do as you prefer" all the way over to "we have an ADR, we lint for it, and reject PRs that don't follow convention".

Adding return types in a lot of places as a public service quality improvement - that I really don't have anything against.

One very minor niggle, part of what led me to ask in the first place about these things, is e.g. whether we should do `JSX.Element | null` as the return type, or just let things go back and forth between `| null` or not based on circumstances changing over time. Also, which JSX? The global one or `React.JSX`? I assume the latter.

I'm not saying no, I'm just gathering information and trying to find the best implementation to land on.

> > > In TypeScript versions before 5.1, you can't use React.ReactNode in one specific case - typing the return type of a component.
> 
> > So does that mean that we need to hold off on that? Especially for external-facing interfaces since we don't force anybody to pick a particularly recent TS version as consumers of our libraries.
> 
> Yes, at least until TypeScript 5.4 is released because of the [TypeScript Releases Policy](https://backstage.io/docs/overview/versioning-policy/#typescript-releases). In the meantime we can either use `JSX.Element` or `React.ReactElement`. I decided to stick with `JSX.Element` since the codebase is already using it. Ideally, I think if we want to change the return type we should do that in another PR.

Alright, see above.

**Comment by schultzp2020** [comment]
@freben 

> Oh there really isn't any hesitancy per se. Everyone is perfectly free to add explicit return types to make things better. This is just me pondering how prescriptive or forcing we should be, on the sliding scale between "do as you prefer" all the way over to "we have an ADR, we lint for it, and reject PRs that don't follow convention".

> Adding return types in a lot of places as a public service quality improvement - that I really don't have anything against.

As you can probably tell, I am in favor of an ADR advocating for explicit return types on the upstream. However, downstream users should be able to decide whether or not to use them. I think this policy would enable the upstream while allowing downstream users to be flexible. 

> One very minor niggle, part of what led me to ask in the first place about these things, is e.g. whether we should do JSX.Element | null as the return type, or just let things go back and forth between | null or not based on circumstances changing over time. Also, which JSX? The global one or React.JSX? I assume the latter.

> I'm not saying no, I'm just gathering information and trying to find the best implementation to land on.

Yeah you are right. I guess I should have looked it up beforehand because I have always thought `| null` seems weird. [Here](https://medium.com/@davidkelley87/stop-using-return-null-in-react-a2ebf08fc9cd) is a good article on what we should use instead. Apparently returning `null` will disconnect a child from the DOM tree whereas `false` will keep child in the DOM tree. Also by returning `false` we will not need to add `| null` to the return type since `JSX.Element` is a superset of `boolean`.

Correct, `React.JSX`.

**Comment by schultzp2020** [comment]
@freben 

> Am I misreading it entirely, or is [this one](https://tech.jotform.com/return-null-vs-return-false-in-react-826d8abcc429) saying the exact opposite in the null/false debate? Haha :) I'm confused.

> React themselves explicitly call out the usage of null, nothing about booleans except as a side effect.

I guess we should read the [official React documentation](https://react.dev/learn/conditional-rendering#conditionally-returning-nothing-with-null) instead of articles haha.

> Well that would anyway mean that the return type would have to be `React.JSX.Element | boolean`, possibly adding `| null` as well for completeness ... Getting a bit wordy.

> I'm slightly in the camp of "is this all really worth it" as you can tell, but I'm keeping an open mind.

I think the return type, when a component potentially returns nothing, should be `React.JSX.Element | null`  because users can easily infer that a component returns `null` using intellisense. Whereas, returning `false` is not mentioned in the official documentation and `false` is not as clear as `null` which literally means `0 value`. If a component cannot return nothing, the return type should be `React.JSX.Element`.

**Comment by Rugvip** [comment]
I'm gonna bring this back to the original issue. I'm all up for a debate return types but that's not really what this particular issue is about 😅 

> According to [ADR006](https://backstage.io/docs/architecture-decisions/adrs-adr006) backstage wants to avoid use of the types `FC` and `SFC`; however, `ComponentType` is a superset of the two types. Therefore, all references to the `ComponentType` type should be removed where possible.

Sorry I'm not following this. Just because a type is expressed in terms of a type that we discourage doesn't mean that the reason for the discouragement applies to the former type? Should we also remove all usages of `React.lazy` because it is expressed in terms of `ComponentType` as well? We need to weight the benefits and drawbacks of using `ComponentType` alone, it doesn't matter how it happens to be defined.

In my view `ComponentType` is the best choice for input types in public API, simply based on its name and it's semantic meaning. It also allows for both class and functional components, which can come in handy if you want error boundaries, but isn't really a big deal nowadays. Now unfortunately `ComponentType` had this side effect of forcing an optional `children` prop and so we haven't used it everywhere. With the React 18 fix I do see any other reasons not to use it at this point.

For output types there's perhaps a bit more debate to be had due to the additional static properties that `ComponentType` has, and it's always nice to keep output types slim. I still feel that it's worth the tradeoff though, as it communicates very clearly what type of value you are receiving. I could imagine a developer of the more junior kind to get confused about `() => JSX.Element` and call it directly as a function, which of course works! ... maybe ... and then might blow up in the future.

**Comment by schultzp2020** [comment]
> Just because a type is expressed in terms of a type that we discourage doesn't mean that the reason for the discouragement applies to the former type? 

If you discourage the usage of said type, should you not also discourage the usage of a superset of said type? When using `ComponentType` it will always be inferred as the `FunctionComponent` type because the majority, if not all, the components are functional components. If that is the case, why aren't we using `FunctionalComponent`? Both of these types have the implicit `children` problem in React 17. Further, is it not best to narrow the type of a variable to the true "type"? If so, this would mean that you cannot get any closer to the true "type" than `(props: Props) => React.JSX.Element`.

> Should we also remove all usages of `React.lazy` because it is expressed in terms of `ComponentType` as well? We need to weight the benefits and drawbacks of using `ComponentType` alone, it doesn't matter how it happens to be defined.

This is a different discussion because we are consuming those types and have the choice of whether or not we want to emit those types downstream. 

> In my view `ComponentType` is the best choice for input types in public API, simply based on its name and it's semantic meaning. 

In the future, it may be best to remove passing components as props and adopt the [Radix](https://www.radix-ui.com/)/[Headless UI](https://headlessui.com/) component architecture model. This is beyond the scope of the PR though. 

> It also allows for both class and functional components, which can come in handy if you want error boundaries, but isn't really a big deal nowadays.

In the future, this will be a non issue because the React Core team are developing an error boundary hook. 

> For output types there's perhaps a bit more debate to be had due to the additional static properties that `ComponentType` has, and it's always nice to keep output types slim. I still feel that it's worth the tradeoff though, as it communicates very clearly what type of value you are receiving. I could imagine a developer of the more junior kind to get confused about `() => JSX.Element` and call it directly as a function, which of course works! ... maybe ... and then might blow up in the future.

I think it is important for React developers to understand the differences between a component and an element at the core level. 

- When it is a function, it is a component. 
- When it is JSX, it is an element. 

When you add the `ComponentType` type it hides the fact that a component is a function at its core. React developers need to understand this construct to build proper composable components. 

These changes would also help the React 18 migration go smoother.

**Comment by Rugvip** [comment]
Yep I think those are some reasonable arguments for the case. Curious if anyone else in the community has any other compelling arguments or preferences, let us know.

**Comment by schultzp2020** [comment]
Can you ping me when we can move forward with this issue?

**Comment by schultzp2020** [comment]
not stale

**Comment by schultzp2020** [comment]
not stale

**Comment by freben** [comment]
Hey, we talked among ourselves and went through the arguments put forward, and decided that usage of this type is not something that we see any significant benefit in discouraging. The react 18 migration is also complete at this point without incident. As such, we'll choose to chose this issue. Thanks for sticking with us through the discussion though!
